### PR TITLE
Speed up MSD.GetNode

### PR DESCRIPTION
**Pull request by benaadams**
Increases performance of MSD.GetNode x2.58 (*edit* for second commit)

Testing with LargeJsonApi app using the POST test scenario x1 warmup and x4 for run

Before 

getnode-before

After 

getnode-after-2

After results above are combined with change https://github.com/aspnet/Common/pull/159

Overall own time before
overall-own-time-before

Overall own time after
overall-own-time-after

I might have got a bit carried away with this PR, don't mind if you think it goes too far... 

Resolves https://github.com/aspnet/Mvc/issues/5362

/cc @rynowak @danroth27 @ajaybhargavb

### CODE REVIEW

**Comment by benaadams** [comment]
With tracing on for call counts

Before
before

After
after

**Comment by benaadams** [comment]
Sped up MSD.MSN.GetNode also

**Comment by rynowak** [comment]
How does the vectorized approach compare to a naive for-loop? 

I have some concerns about the maintainability of the vector approach. I'll discuss with some folks here. I think we have a different set of criteria than the core HTTP parsing logic in Kestrel :laughing: For one, we own the definition of these keys, it's not a standard, so it's more likely to evolve in the future.

File: src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelStateDictionary.cs [hunk 1]
```diff
@@ -806,26 +1128,35 @@ public void Reset()
...
-                }
-
-                var index = BinarySearch(subKey);
                 ModelStateNode modelStateNode = null;
-                if (index >= 0)
-                {
-                    modelStateNode = ChildNodes[index];
-                }
-                else if (createIfNotExists)
+                if (subKey.Length != 0)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelStateDictionary.cs]: What's the strategy here? Do you find that `ChildNodes ==` null is the common case and we can avoid the binary search when that happens?

**Comment by benaadams** [comment]
Updated result captures with second commit results

**Comment by benaadams** [comment]
> How does the vectorized approach compare to a naive for-loop?

Just need to comment out lines 547 - 747

Original
before
Non-vector
after-nonvector
Vector
after

**Comment by benaadams** [comment]
The Vector path will get faster as the keys get longer; but I don't know how common that is ?

**Comment by benaadams** [comment]
Overall own time before
overall-own-time-before

Overall own time after
overall-own-time-after

**Comment by benaadams** [comment]
Overall own time before
overall-own-time-before

Overall own time after
overall-own-time-after

The three changes seem to be disproportionately significant for this scenario

**Comment by rynowak** [comment]
Comparing your last two data points, a lot of stuff you didn't change is varying by ~500 ms across these two runs. I think we'll have to have someone on the team take a more thorough look at measuring in a more stable way than just profiling.

=======================

That said, I think we have a high bar for starting to put vectors into MVC :laughing: - we'll also need to look into reducing the number of calls we're making into `GetNode`. https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Validation/ValidationVisitor.cs#L155

The visitor was written back when `ModelStateDictionary` was backed by a dictionary, now that it's a tree we can be much more efficient by traversing the nodes directly.

**Comment by benaadams** [comment]
Less exciting version in https://github.com/aspnet/Mvc/pull/5389

**Comment by benaadams** [comment]
Closing, replaced with https://github.com/aspnet/Mvc/pull/5389 which now performs better

**Pull request closed without merging**